File: src/Compile.fs
module Compile

exception private LexError of string
exception private ParseError of string
exception private ResolveError of string
exception private LoopLabelError of string

let private compileInner (config: Settings.CompilerConfig) (stage: Settings.Stage) (optimizations: Settings.Optimizations) (src_file: string) (source: string) =
    let counter = UniqueIds.initialCounter
    // Lex it
    let tokens =
        match Lexer.lex source with
        | Ok toks -> toks
        | Error msg -> raise (LexError msg)
    if stage = Settings.Lex then ()
    else
        let ast =
            match Parse.parse tokens with
            | Ok a -> a
            | Error msg -> raise (ParseError msg)
        if stage = Settings.Parse then
            printf "%A" ast
        else
            // Semantic analysis has three steps:
            // 1. resolve identifiers
            let counter', resolved_ast =
                match Resolve.resolve counter ast with
                | Ok r -> r
                | Error msg -> raise (ResolveError msg)
            // 2. annotate loops and break/continue statements
            let counter'', annotated_ast =
                match Label_loops.labelLoops counter' resolved_ast with
                | Ok r -> r
                | Error msg -> raise (LoopLabelError msg)
            // 3. typecheck definitions and uses of functions and variables
            // Sync shared counter before typecheck (it uses Symbols.addString which uses the shared counter)
            UniqueIds.sharedCounter <- counter''
            let typed_ast = Typecheck.typecheck annotated_ast
            let counter_after_typecheck = UniqueIds.sharedCounter
            if stage = Settings.Validate then ()
            else
                // Convert the AST to TACKY
                let counter''', tacky = TackyGen.gen counter_after_typecheck typed_ast
                // print to file (src filename with .debug.tacky extension) if debug is
                // enabled
                let counter'''' = Tacky_print.debugPrintTacky config.Debug counter''' src_file tacky
                // optimize it!
                let optimized_tacky = Optimize.optimize optimizations src_file tacky
                if stage = Settings.Tacky then ()
                else
                    // start by getting all aliased vars, we'll need them for register
                    // allocation
                    let aliased_vars = Address_taken.analyzeProgram optimized_tacky
                    // Assembly generation has three steps:
                    // 1. convert TACKY to assembly
                    let _counter5, asm_ast = Codegen.gen counter'''' optimized_tacky
                    // Sync shared counters for modules that still use them (e.g. Cfg.printGraphviz)
                    UniqueIds.sharedCounter <- _counter5
                    Cfg.setCounter _counter5
                    // print pre-pseudoreg-allocation assembly if debug enabled
                    if config.Debug then
                        let prealloc_filename =
                            System.IO.Path.ChangeExtension(src_file, null) + ".prealloc.debug.s"
                        Emit.emit config.Platform prealloc_filename asm_ast
                    // replace remaining pseudoregisters with Data/Stack operands
                    let asm_ast1 = Regalloc.allocateRegisters config.Debug aliased_vars asm_ast
                    if config.Debug then
                        let postalloc_filename =
                            System.IO.Path.ChangeExtension(src_file, null) + ".postalloc.debug.s"
                        Emit.emit config.Platform postalloc_filename asm_ast
                    let asm_ast2 = ReplacePseudos.replacePseudos asm_ast1
                    // fix up instructions
                    let asm_ast3 = InstructionFixup.fixupProgram asm_ast2
                    if stage = Settings.Codegen then ()
                    else
                        let asm_filename = System.IO.Path.ChangeExtension(src_file, ".s")
                        Emit.emit config.Platform asm_filename asm_ast3

let compile (config: Settings.CompilerConfig) (stage: Settings.Stage) (optimizations: Settings.Optimizations) (src_file: string) (source: string) : Result<unit, CompilerError.CompilerError> =
    try
        Ok (compileInner config stage optimizations src_file source)
    with
    | LexError msg -> Error (CompilerError.LexError msg)
    | ParseError msg -> Error (CompilerError.ParseError msg)
    | ResolveError msg -> Error (CompilerError.ResolveError msg)
    | LoopLabelError msg -> Error (CompilerError.LoopLabelError msg)
    | Failure msg when not (msg.StartsWith("Internal error")) ->
        Error (CompilerError.TypeError msg)

File: src/Compile.rs
//! The compiler's source-to-assembly pipeline.
#![allow(non_snake_case)]

use std::any::Any;
use std::path::Path;
use std::panic::{catch_unwind, resume_unwind, AssertUnwindSafe};

use super::{AddressTaken, Cfg, Codegen, CompilerError::CompilerError, Emit,
    InstructionFixup, LabelLoops, Lex, Optimize, Parse, Regalloc, ReplacePseudos,
    Resolve, Settings::{CompilerConfig, Optimizations, Stage}, TackyGen, TackyPrint,
    Typecheck, UniqueIds};

#[derive(Debug)]
enum CompileFailure {
    Lex(String),
    Parse(String),
    Driver(String),
}

fn compileInner(config: &CompilerConfig, stage: Stage, optimizations: &Optimizations,
                src_file: &Path, source: &str) -> Result<(), CompileFailure> {
    let counter = UniqueIds::INITIAL_COUNTER;
    let tokens = Lex::lex(source).map_err(CompileFailure::Lex)?;
    if stage == Stage::Lex { return Ok(()); }

    let ast = Parse::parse(tokens).map_err(CompileFailure::Parse)?;
    if stage == Stage::Parse {
        print!("{ast:?}");
        return Ok(());
    }

    let (counter, resolved_ast) = Resolve::resolve(counter, ast);
    let (counter, annotated_ast) = LabelLoops::labelLoops(counter, resolved_ast);
    UniqueIds::set_shared_counter(counter);
    let typed_ast = Typecheck::typecheck(annotated_ast);
    let counter = UniqueIds::shared_counter();
    if stage == Stage::Validate { return Ok(()); }

    let (counter, tacky) = TackyGen::gen(counter, typed_ast);
    let counter = TackyPrint::debugPrintTacky(config.Debug, counter, src_file, &tacky)
        .map_err(|error| CompileFailure::Driver(error.to_string()))?;
    let optimized_tacky = Optimize::optimize(optimizations, src_file, tacky);
    if stage == Stage::Tacky { return Ok(()); }

    let aliased_vars = AddressTaken::analyzeProgram(&optimized_tacky);
    let (counter, asm_ast) = Codegen::gen(counter, optimized_tacky);
    UniqueIds::set_shared_counter(counter);
    Cfg::setCounter(counter);
    if config.Debug {
        Emit::emit(config.Platform, src_file.with_extension("prealloc.debug.s"), &asm_ast)
            .map_err(|error| CompileFailure::Driver(error.to_string()))?;
    }
    let asm_ast1 = Regalloc::allocateRegisters(config.Debug, &aliased_vars, asm_ast.clone());
    if config.Debug {
        // This deliberately matches the original, which emits the pre-allocation
        // tree to both debug files.
        Emit::emit(config.Platform, src_file.with_extension("postalloc.debug.s"), &asm_ast)
            .map_err(|error| CompileFailure::Driver(error.to_string()))?;
    }
    let asm_ast2 = ReplacePseudos::replacePseudos(asm_ast1);
    let asm_ast3 = InstructionFixup::fixupProgram(asm_ast2);
    if stage != Stage::Codegen {
        Emit::emit(config.Platform, src_file.with_extension("s"), &asm_ast3)
            .map_err(|error| CompileFailure::Driver(error.to_string()))?;
    }
    Ok(())
}

fn panic_message(payload: &(dyn Any + Send)) -> Option<&str> {
    payload.downcast_ref::<String>().map(String::as_str)
        .or_else(|| payload.downcast_ref::<&'static str>().copied())
}

/// Compile one already-loaded translation unit through the requested stage.
/// Internal compiler panics remain panics; user-facing validation failures are
/// converted to `TypeError`, matching the F# `Failure` handler.
pub fn compile(config: &CompilerConfig, stage: Stage, optimizations: &Optimizations,
               src_file: impl AsRef<Path>, source: &str) -> Result<(), CompilerError> {
    match catch_unwind(AssertUnwindSafe(|| {
        compileInner(config, stage, optimizations, src_file.as_ref(), source)
    })) {
        Ok(Ok(())) => Ok(()),
        Ok(Err(CompileFailure::Lex(message))) => Err(CompilerError::LexError(message)),
        Ok(Err(CompileFailure::Parse(message))) => Err(CompilerError::ParseError(message)),
        Ok(Err(CompileFailure::Driver(message))) => Err(CompilerError::DriverError(message)),
        Err(payload) => match panic_message(payload.as_ref()) {
            Some(message) if !message.starts_with("Internal error") =>
                Err(CompilerError::TypeError(message.to_owned())),
            _ => resume_unwind(payload),
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use super::super::Settings::Target;

    fn config() -> CompilerConfig { CompilerConfig { Debug: false, Platform: Target::Linux } }

    #[test]
    fn reports_lex_and_parse_errors() {
        let options = Optimizations::default();
        assert!(matches!(compile(&config(), Stage::Lex, &options, "bad.c", "@"),
                         Err(CompilerError::LexError(_))));
        assert!(matches!(compile(&config(), Stage::Parse, &options, "bad.c", "int ;"),
                         Err(CompilerError::ParseError(_))));
    }

    #[test]
    fn honors_early_pipeline_stages() {
        let options = Optimizations::default();
        assert_eq!(compile(&config(), Stage::Lex, &options, "ok.c", "int main(void) { return 0; }"), Ok(()));
    }
}
